### ISSUE TITLE
ClassMap

### ISSUE THREAD

**Issue by MichaelTheBeast** [roadrunnerlenny/etlbox#5]
Hi

I've recently started an ETL project and I'm looking into the possibilities of ETLbox.
So far it is looking very good, but I'm having some difficulties with the following scenario:

The CsvSource output is a string array, which gives the RowTransformation this form:
`RowTransformation<string[], MyData> row = new RowTransformation<string[], MyData>
(
    input => new MyData() 
    { Col1 = input[0], 
      Col2 = input[1] }
);`

With 50-60 fields in the input-file, a string array is hard to manage.

The alternative CsvSource<TOutput> requires the TOutput to precisely match the Csv-file.
It could be acceptable, but when CsvHelper already has the ability to use a ClassMap and it is possible to autogenerate the class and classmap, why not implement it? Or am I missing something?

Also, I have often experienced that the Csv-file is flattened, so a single file ends up in multiple database tables.
Is that a possibility in EtlBox?

Regards
Michael

**Comment by roadrunnerlenny** [comment]
Hey Michael,

Thanks for your interest in ETLBox! 

You are absolutely right, if you would have an input file with like 50 fields, a string array would be hard to manage. And I agree, the generic `CsvSource` with `TOutput` would require the properties names exactly to match the field names in the csv file, there is currently no other mapping functionality implemented.
I just had a look at the ClassMap in the CsvHelper API - and yes, it would be a good idea to use the same mapping features as provided by the api. 
I will have a deeper look at at the ClassMap in CsvHelper, and I can hopefully incorporate it into the current `CsvSource` in the next couple of days, so that you will be able to use the same features like in the CsvHelper class. I will give you update on the progress here.
But I already have a question here: I see that you can do redefine the mapping of columns in CSV file to the property in the class, but I didn't see anything about autogenerating classes at first glance. What do you mean with this?

Regarding your question about flattened (or de-normalized) csv files: 
The approach to normalize these structures would be to use a Multicast.
A multicast would duplicate your data rows and give you two outputs - each output would point to a different table and would only process the columns you are interested in. 
More infos about the multicast can be found here:
https://etlbox.net/api/ALE.ETLBox.DataFlow.Multicast-1.html
https://github.com/roadrunnerlenny/etlbox/blob/83d1d3b8eec888e4e7e41fad9fca7961aad52309/ETLBoxTest/src/DataFlow/TestDataFlowMulticast.cs#L38

Does that help with your second question?

**Comment by MichaelTheBeast** [comment]
Hi Roadrunner

Thank you for the quick reply.
Here is a link to the autogenerator:
[https://toolslick.com/generation/code/class-from-csv](url)
It generates a Poco from the given Csv-file AND a classmap to use with CsvHelper :)

Regarding the normalization from a single Csv-file to multiple entities/tables...
Multicast might be the right way to go; I was not aware that I can choose which fields/columns to process.

Perhaps I should elaborate :)
I have about 10 different companies who is providing me with Csv-files (some give 2 files an other gives me 6 files).
I have created a staging database with about 15 tables, for each company.
So for each company I read the Csv-files, converts them to Poco's (to make the transformation easier).
The Poco's must be transformed in various ways and mapped to the right place in my database-structure.

After this the data has to be synchronized into a consolidation database that contains the data from all companies.

Back to the Multicast...
Multicast has only one input so the input has to be prepared but how?
The separation into multiple objects is done in the RowTransformation or is it?
 
```
var firstEntityRow = new RowTransformation<MyPoco, FirstEntity>(
              input => new FirstEntity()
              {
                Id = Guid.NewGuid(),
                Name = input.PocoName
                // and so on
              });
```

**Comment by roadrunnerlenny** [comment]
Hey Michael, 
regarding the class map functionality from CSVHelper, I was able to add this to the CSVSource as well. Actually, it simplified the code - thanks for the good advice!

Here is the test case that makes use of the mapping attributes:
https://github.com/roadrunnerlenny/etlbox/blob/17e5d0b6cec42b526563d4b1b8ad5c62b4f202e8/ETLBoxTest/src/DataFlow/TestDataFlowCSVSource.cs#L144-L166

And here is an example that makes use of the classmap feature:
https://github.com/roadrunnerlenny/etlbox/blob/0b8e8fb44567ee2db8476934ce43da2bdafb6e32/ETLBoxTest/src/DataFlowExamples/TestIssue5_CSVSourceWithClassMap.cs#L29-L61

**Comment by roadrunnerlenny** [comment]
Regarding the Multicast:
I am not quite sure if I fully understood your issue here. 

So the idea of a multicast is basically to have one input, but two outputs (basically duplicates). 
Here's the idea how you would use that Multicast in your scenario:
First of all, you convert your CSV into a POCO
```CSVSource<POCO> source = new CSVSource<POCO>();```
Now let's you want to have particular columns of the CSV in table1 and some in table2. 
You would now add the Multicast, (still of the type POCO)
```Multicast<POCO> mc = new Multicast<POCO>();```
The Multicast can now be linked to two target. Now we use 2 RowTransformations to transform our POCO type into the type we need for table1 and table2.
```RowTransformation<POCO, TTable1> rt1 = new RowTransformation<POCO, TTable1>(transFunc1);
RowTransformation<POCO, TTable2> rt2 = new RowTransformation<POCO, TTable2>(transFunc2);```

Now we need the DBDestination for each table
```DBDestination<TTable1> dest1 = new DBDestination<TTable1>("table1");
DBDestination<TTable2> dest2 = new DBDestination<TTable2>("table2");```

Finally, we tie this all together:
```source.LinkTo(mc);
mc.LinkTo(rt1);
mc.LinkTo(rt2);
rt1.LinkTo(dest1);
rt2.LinkTo(dest2);
```
And then start the whole thing
```
source.Execute();
dest1.Wait();
dest2.Wait();
```

Does that help?

**Comment by roadrunnerlenny** [comment]
One last comment: The support of CSVHelper classmap / CSVHelper mapping attributes for the CSVSource is currently in the dev branch. If it helps you out, I will try to create a new release (version 1.1.2) in the next couple of days so that the changes are available within nuget.

**Comment by MichaelTheBeast** [comment]
Hey Roadrunner

Good to hear about the new mapping functionality :)

I will try your suggestion with the MultiCast as soon as possible...
My confusion was where to fit the RowTransformation, so your example is quite interesting :)

**Comment by MichaelTheBeast** [comment]
Unfortunately something is not right...
I have created 2 tables with 4 columns in each (more than needed for this test):
```
// Creates basic TableDefinition with 4 columns
private TableDefinition CreateTable(string tablename)
    {
      var def = new TableDefinition(tablename, new List<TableColumn>
      {
        new TableColumn("Col1", "nvarchar(100)", true),
        new TableColumn("Col2", "nvarchar(100)", true),
        new TableColumn("Col3", "nvarchar(100)", true),
        new TableColumn("Col4", "nvarchar(100)", true)
      });
      def.CreateTable();
      return def;
    }
var tableDestination1 = this.CreateTable("Table1");
var tableDestination2 = this.CreateTable("Table2");
```
Then I added the CsvSource:
```
var source = new CSVSource<TestPoco>("test.csv"){ Delimiter = ";" };
```
The test.csv also contains 4 columns:
```
Value1;Value2;Value3;Value4
one;two;three;four
```
And the TestPoco to receive the data with matching column names:
```
internal class TestPoco
  { public string Value1 { get; set; }
    public string Value2 { get; set; }
    public string Value3 { get; set; }
    public string Value4 { get; set; }  }
```
And the transformations:
```
var row1 = new RowTransformation<TestPoco, TestEntity1>(input =>{
        return new TestEntity1
        { Col1 = input.Value1,
          Col3 = input.Value3 };});
var row2 = new RowTransformation<TestPoco, TestEntity2>(input =>{
          return new TestEntity2
          { Col2 = input.Value2,
            Col4 = input.Value4 };});
```
With the 2 outputs defined as:
```
internal class TestEntity1
  { public string Col1 { get; set; }
    public string Col3 { get; set; } }

  internal class TestEntity2
  { public string Col2 { get; set; }
    public string Col4 { get; set; } }
```
And putting it all together:
```
var multicast = new Multicast<TestPoco>();
var destination1 = new DBDestination<TestEntity1>("Table1");
var destination2 = new DBDestination<TestEntity2>("Table2");

multicast.LinkTo(row1);
multicast.LinkTo(row2);

row1.LinkTo(destination1);
row2.LinkTo(destination2);

source.Execute();
destination1.Wait();
destination2.Wait();
```
**The expected result would be:**
Table1: a single row with values 'one' and 'three' in Col1 and Col3
Table2: a single row with values 'two' and 'four' in Col2 and Col4

but when the program ends with **destination1.Wait()** where it waits forever.
and before you ask...
yes the tables was created :)

Any ideas?

**Comment by roadrunnerlenny** [comment]
When putting it all together, you forget to link the source to the Multicast.

```C#
source.LinkTo(multicast);
```

**Comment by MichaelTheBeast** [comment]
Aaaah yes and now the data is coming through to the database :)

Next challenge...
Table1:
![image](https://user-images.githubusercontent.com/48888742/55151379-6bd60d80-514e-11e9-875d-89cc36bc69a5.png)
Table2:
![image](https://user-images.githubusercontent.com/48888742/55151427-85775500-514e-11e9-8bd8-cd3306ecc36e.png)

Is there any way of debugging the process? or do you have a suggestion?

**Comment by roadrunnerlenny** [comment]
Hey Michael,

Thanks for your patience. I added your code as a Test to the project in order to reproduce the error. (https://github.com/roadrunnerlenny/etlbox/blob/master/ETLBoxTest/src/DataFlowExamples/TestIssue5_CSVSourceWithClassMap.cs)

I could reproduce a similar behaviour (actually, I also had values in Col2). But the `DBDestination`did not work as expected - you actually found a bug here. Everything would have worked if Table1 would have only Col1 and Col3, and Table2 Col2 and Col4. (This was the way I designed the tests, so I didn't encounter this bug before)

The problem with the `DBDestination` is that it does have problems with mapping column names with the property names. 

I already create a fix for that and pushed it into the repo (dev & master). Also, I release version 1.1.2 including the fix and uploaded the new version to nuget: https://www.nuget.org/packages/ETLBox/1.1.2

Could you update your project to reference the latest version (now 1.1.2) and try again? It should work now.

Just a quick warning: When you create the CSVSource, the code for changing the configuration of the CSVHelper is now:
```
var source = new CSVSource<TestPoco>("test.csv"){ 
  new CsvHelper.Configuration.Configuration() { Delimiter = ";" }
};
```

**Comment by MichaelTheBeast** [comment]
Sorry; no visible changes:
![image](https://user-images.githubusercontent.com/48888742/55229759-2e40b580-521e-11e9-974b-2f4a5694e2af.png)
except this:
```
var source = new CSVSource<TestPoco>(file);
source.Configuration.Delimiter = ";";
```

**Comment by roadrunnerlenny** [comment]
Hey Michael,

that is strange. I created a test case which basically does the things as described above:
https://github.com/roadrunnerlenny/etlbox/blob/516854d8540057706a595c90ee3694666cbac4fd/ETLBoxTest/src/DataFlowExamples/TestIssue5_CSVSourceWithClassMap.cs#L67-L141

When I run the test, this is what I get in the tables:
<img width="242" alt="grafik" src="https://user-images.githubusercontent.com/6517572/55257867-7c2add00-5261-11e9-87aa-be007c6086e5.png">

Which actually is what I expect. Perhaps you did something different in your code? 
Could you have a look a the test and compare at with your code, and if you can't spot a difference, could you copy/paste your code here? Then I can have a look at it.

**Comment by MichaelTheBeast** [comment]
Hey Roadrunner

I've got an explanation now...
Apparently it is required to specify the schema, I've assumed that dbo was default.

So onward to the next issue :)
As a proof of content I've tried to add file-logging to the test, but nothing happens and no errors.
I've created a nlog.config like this:
```
<?xml version="1.0" encoding="utf-8"?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xsi:schemaLocation="NLog NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      autoReload="true">
  <rules>
    <logger name="*" minlevel="Trace" writeTo="file, file2" />
  </rules>
  <targets>
    <target name="console" xsi:type="Console" />
    <target name="file" xsi:type="File" layout="${date}|${level:uppercase=true}|${message} ${exception}|${logger}|${all-event-properties}" 
            fileName="$D:\logfile.txt" />
    <target name="file2" xsi:type="File" fileName="${basedir}/Test.log"
            layout="${date}|${level:uppercase=true}|${message} ${exception}|${logger}|${all-event-properties}"/>
  </targets>
</nlog>
```
I've added
`LogTask.Debug("Test startet");`
And
`var logTask = new LogTask("New LogTask...");`
just to get any response, but nothing (later I want to add a SQL-log).
Can you provide more info how the logging is enabled (I'm new to nlog, but it looks pretty straightforward).

Cheers

**Comment by roadrunnerlenny** [comment]
Hey Michael,

that by default the dbo schema is not used is actually a bug that I will fix in the next release.  Thanks for finding out!
I had a look at the nlog configuration, I think there are two issues here.
First, you should add the the internal exception log from nlog to the nlog tag, so that it looks like this:
```
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xsi:schemaLocation="NLog NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      autoReload="true"
      throwConfigExceptions="true"
      internalLogFile="console-example-internal.log"
      internalLogLevel="Info">
```
You will then find a file console-example-internal.log that contains exceptions in your log file.
When I tried to log into c:\Logfile.exe, i got an exception: "Exception: System.UnauthorizedAccessException: Access to the path 'C:\logfile.txt' is denied.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)"

I think you will also have a similiar error message due to restrictions with security in .net core (it probably behaves different in .net)
But the logging into  "${basedir}/Test.log" did work on my machine. 

Make sure that the nlog.config is copied into the output directory. In Visual Studio 2017, right-click on the nlog.config file -> Properties -> Copy to Output Directoy -> Set to "Always" of "Copy if newer".
Alternatively, if you use another IDE, edit the project file and add 
```
<ItemGroup>
    <None Update="nlog.config">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
...
```

These modifications should help you to enable logging. Let me know if you have any problems.

**Comment by MichaelTheBeast** [comment]
Thank you Roadrunner

I have logfiles now :)
It was the internal log that did it.

**Comment by MichaelTheBeast** [comment]
Hi again

What about handling relations...
I have a one-to-many relation so I'm thinking of having 2 RowTransformations (1 for the parent and 1 for the child):
```
var row1 = new RowTransformation<Poco, Parent>(in =>
{
    return new Parent
   {
     Id = Guid.NewGuid(),
     Name = in.Name
     RefParentId = in.Id // the originally id from the file (could be used to create the relation).
   };
 });
var row2 = new RowTransformation<Poco, Child>(in =>
 {
    return new Child
    {
     Id = Guid.NewGuid(),
     Name = in.Name,
     RefChildId = in.Id,
     RefParentId = in.ParentId,
     Parent = // find parent-Entity via in.ParentId? OR
     ParentId = // find the parentId GUID
   };
});
```

**Comment by roadrunnerlenny** [comment]
Hey Michael,

I think this is a typcial ETL issue you have here - something where also SSIS (an ETL tool from Microsoft) would give several approaches to solve this.

First of all, you could think of "staging" all your data first - just store the data in a table, where you have a primary key with identity auto increment defined. (It should be of database type INT, if you use a UNIQUEIDENTIFER you can get other issues, but there are probably a lot of discussions in the web about this topic)
Once you staged all your data, you can have a second dataflow - there you can use the `Lookup` component to find foreign keys relations.

But perhaps something which helps you more is the `BlockTransformation`. A `RowTransformation`just processes all data row by row, without stopping the flow of incoming data. It is not possible here to find relations to other rows, because you can't determine if the row you are looking for is already in the pipe or still in the source waiting to be loaded.  This is were the `BlockTransformation` comes in  - it will "block" the complete flow until all rows are loaded into the component (make sure you have enough memory if you load a big amount of data). This way, you can be sure that all rows are avaible in the component (so all data from the source is loaded, and no data was forwarded in the flow). Now you can describe your own lookup. 
See this example test code:
https://github.com/roadrunnerlenny/etlbox/blob/3b63b12cef6b67405ec2a39504817c6450824268/ETLBoxTest/src/DataFlow/TestDataFlowBlockTransformation.cs#L28-L74

You basically get a list of all Rows - there you can define your own lookup definition,e.g. in the POCO you define a Property Parent and Child, and then you define your own algorithm to populate them in the list. You can use all .NET Core List-operation for this, and everything happens in the memory, so it should work quite fast. 

Does that make sense to you?

**Comment by MichaelTheBeast** [comment]
Very intriguing; that might be useful.

But before I dive into that:
I have removed the constraints between the tables in the database (a staging database)...
But is it possible to avoid duplicate entries (the file contains both parent and children)?
```
var row1 = new RowTransformation<Poco, Parent>(in =>
{
    return new Parent // if not already added!
   {
     Id = in.Id,
     Name = in.Name
   };
 });
```

**Comment by roadrunnerlenny** [comment]
Hey Michael,

the `BlockTransformation` could help you there as well as you would have a List containing all items, and you could filter out duplicates in there and return the new List. 

If you really want to do it in your flow without stopping it (meaning to wait for all data to be loaded and then change the whole List) you can use the `RowTransformation` in combination with the predicates. A predicate is a filter expression that can be used when linking two components (basically it's a filter).

This would look like this:
```C#
List<int> IDs = new List<int>(); //at the end of the flow, this list will contain all IDs of your source
var row1 = new RowTransformation<Poco, Poco>(in =>
{
    isDuplicate = false;
    if (IDs.Contains(in.Id))
     isDuplicate = true;
   else 
       IDs.Add(in.ID);
    return new Poco // if not already added!
   {
     Id = in.Id,
     Name = in.Name,
    IsDuplicate = isDuplicate
   };
 });

var dest = new DBDestination<Poco>("table1");
row1.LinkTo(dest, in => in.IsDuplicate == false);
```

I haven't tested this example beforehand, so perhaps you need to modify it a little bit. But I hope you get the idea behind it.

**Comment by MichaelTheBeast** [comment]
Hey Roadrunner

Good plan, didn't work ;)

Isn't the BlockTransformation "just" creating a new version of the Poco itself?
```
var block = new BlockTransformation<Poco>( 
     inputData => { 
         return inputData.Select( row => new Poco() { Col1 = row.Col1, Col2 = 3 }).ToList(); }); 
```
The Rowtransformation (I have removed the second RowTransformation for simplicity, but kept the MultiCast):
```
var cabinIds = new List<string>();
var row1 = new RowTransformation<Cabin, Ship>(input =>
   {
      var isDuplicate = false;
      if (cabinIds.Contains(input.SHIPCD))  
        isDuplicate = true;
      else
        cabinIds.Add(input.SHIPCD);
      return new Ship
      {
                Id = input.SHIPCD,
                Name = input.SHIPNAME,
                RefShipId = input.SHIPCD,
                IsDuplicate = isDuplicate
      }; });
```
And the plumbing:
```
var source = new CSVSource<Cabin>(file){ Configuration = new Configuration(){Delimiter = ";"}};
source.Configuration.RegisterClassMap<CabinClassMap>();
var multicast = new Multicast<Cabin>();
source.LinkTo(multicast);
multicast.LinkTo(row1);
row1.LinkTo(destination1, input => input.isDuplicate == false);
source.Execute();
destination1.Wait();
```

But the destination never returns.
I wonder if it's something crazy like 'IsDuplicate' has to exist in the database-table.

**Comment by roadrunnerlenny** [comment]
Hey Michael,

I did write a test that contained the code from my previous answer. I had the same behavior - the destination never returned. This is because the destination does wait for the completion message from the source - and the completion message is only received if all items in the pipe where processed. 
But because of the filter only some of the rows are the processed - the rest is "stuck" in the pipe and so the Dataflow will never finish.

To get this to work you need to add a multicast. The multicast will send the "good" rows (these which are not duplicates) into the DB table and the other into a `CustomSource`. The CustomSource does just need to receive the "bad" data to make it work. In my example code it will also print out a Log message informing about the duplicate. 

So this is how the duplicate check with the `RowTransformation` would work: (This time tested code)

https://github.com/roadrunnerlenny/etlbox/blob/bdf670ad5a4a8ec7eea0239cdb9398a02b6edcb5/ETLBoxTest/src/DataFlowExamples/TestIssue5_DuplicateCheck.cs#L32-L84

Regarding the `BlockTransformation`: The BlockTransformation is not creating a new version of the Poco itself. It gives you a List of all Pocos from your source (which are basically all rows in your csv file), and you can modify this List according your needs. In your case, this would be that you would remove all duplicates from the List and then just return the modified List. I created an example, whereby I use Linq to remove the duplicates from the List:

https://github.com/roadrunnerlenny/etlbox/blob/bdf670ad5a4a8ec7eea0239cdb9398a02b6edcb5/ETLBoxTest/src/DataFlowExamples/TestIssue5_DuplicateCheck.cs#L87-L113

Hope this helps you.

**Comment by roadrunnerlenny** [comment]
Hey Michael, 

were you able to make things work?

**Comment by MichaelTheBeast** [comment]
Hey Roadrunner

Sorry for the late reply...
I fear it's evolving out of control.

An example of what I hope to achieve:

- 1 Csv-file containing 100.000 records in 40 columns.
- Transformed into a List<Poco> with 100.000 instances of the Poco with 40 properties.
- The List<Poco> transformed into 4 Entities with different number of instances and properties.
- And then persisting the 4 collections of Entities to the database.

So to achieve this I'd have to do something like:

- Create a Mulitcast
- Create a Source
- Link Source to Multicast
- Create 4 BlockTransformations
- Link Multicast to the 4 BlockTransformations
- Create 4 destinations and Link the 4 BlockTransformations to the destinations.

Am I right?

**Comment by roadrunnerlenny** [comment]
Hey Michael,

this Plan Sounds good. The only thing I see here is that you would need 3 multicasts instead of one. 

In the current version a multicast has one input and two outputs. So if you would like to have a multicast with one input and 4 outputs, you basically have to take one multicast and connect each output to another multicast. This way you get 4 outputs with the same object/data.

I will add this on my todo-list to have a multicast component with a random number of outputs.

But the rest of what you wrote sounds like it will work.

It looks like a lot of components, but please keep in mind that it is an ETL framework and designed for a high throughput  with a big amount of data. So this should not only work with 100.000, but also with 1 Million or 10 million rows (if you have enough memory which is needed for the BlockTransformations)

It would be great if me let me know if this worked out for you or not.

**Comment by roadrunnerlenny** [comment]
Hey Michael, 

I just edited my previous comment. The current `Multicast` implementation does allow to connect it with as many targets as you want. Therefore, you will need only 1 Multicast which is linked 4 times to your destinations.
I wasn't aware of this because the implementation heavily depends on the `BroadcastBlock` from the TPL.Dataflow library. I just added a test and it works to connect the Multicast with more than 2 targets.
 
Just to let you know - I just released version 1.2.0. It does contain some minor changes - perhaps interesting for you is the possibility to add the attribute`[ColumnMap("columnname")]` to the property of a POCO used in a DBDestination. It will then map the property to the particual column.
E.g. 
```
public class MySimpleRow {
    [ColumnMap("Value1")]
    public string Col1 { get; set; }
    public int Value2 { get; set; }
}
```
Would write the value of Col1 into the database column with the name Value1.
